src/libraries/System.Linq/src/System/Linq/Flatten.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        /// <summary>
+        /// Flattens an enumerable of enumerables into a single concatenated enumerable.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of the input sequences.</typeparam>
+        /// <param name="sources">The source of enumerables to flatten.</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> enumerates all inner enumerates in a concatenated view.</returns>
+        public static IEnumerable<TSource> Flatten<TSource>(this IEnumerable<IEnumerable<TSource>> sources)
+        {
+            if (sources == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(sources));
+            }
+
+            return FlattenIterator(sources);
```

Comment:
Did you consider just doing:
```C#
return source.SelectMany(e => e);
```
rather than implementing another iterator?

Reply:
I did, ultimately it's tradeoff between a closure allocation and a few virtual calls. I should point out that the simple `SelectMany` implementation does inherit from `Iterator<T>`, but ultimately I didn't find any of the speed optimizations offered in this particular case to be useful in practice.

Reply:
> I did, ultimately it's tradeoff between this and a closure allocation and a few virtual calls

There's no closure allocation with `sources.SelectMany(e => e)`.  There is one delegate per T allocated for the process.  And an extra delegate invocation per enumerable.  From my perspective, I'd rather have less code to maintain and a smaller binary.  Up to you, though.